### DISCUSSION TITLE
Coding standards???

### DISCUSSION THREAD

**Discussion by codemasher** [chillerlan/php-qrcode#60] [README]
# What coding standards do you follow? Why don't you adhere to PSR-1 or PSR-12?

First off: i don't enforce any coding standard. If you'd like to contribute code to this project (or any other projects of mine), you are welcome to do so - no matter what code style your contribution is written in. I can always run it through a formatter so that it aligns with the project's style.
That said, this doesn't mean that there aren't any rules at all - most importantly your code should be self-describing and/or be commented so that a reader understands what's happening. I'll lay out the general rules that *i* follow later.

**Why don't you adhere to PSR-1 or PSR-12?**

I do actually follow a lot of the proposals in these PSRs, however, i see several things in both of these as messy and inconsistent and opinionated (as all coding standards), making code harder to read and edit. I mean, just [look at it](https://www.php-fig.org/psr/psr-12/). The working group had a chance to fix things and make it future proof but somehow blew it - not to mention PHP8 (attributes, constructor property promotion etc.) isn't even considered in the recent PSR-12. 
One of the biggest gripes is for sure the choice of [space indentation over tabs](https://lea.verou.me/2012/01/why-tabs-are-clearly-superior/), which is not only silly now in 2021 when everyone and their grandma is using a smart IDE that allows to adjust tab sizes, but it also adds overhead and makes the code a LOT harder to edit (try copy pasting spaced vs tabbed code for example).
The next thing is the inconsistent bracing that makes code harder to read and adds a lot of unnecessary vertical space (in case of classes and methods), then again closing and opening braces on the same line in control structures (probably to make up for the wasted space elsewhere? We'll never know). I know, i know the latter *might* come in handy in case you want to outcomment something for whatever obscure reason, but i've never seen that case out in the wild, nor did i ever consider to use it on my own.
A thing that's completely missed (yet somehow expected) is *properly naming things* so it seems completely fine to name your class that does a very specific thing `MyCoolClass` that has a method `doStuff()` as long as it follows the recommended naming scheme...

**So what are your rules then?**

What follows are the general rules i (try to) adhere to and that are set in my global PhpStorm settings (which [i may add to the repo](https://github.com/chillerlan/php-qrcode/commit/0c1fdb92c345ce4d87bc4f297e2517073b1af796) at some point).

  - Tabs for indentation (tab size 4!), spaces for alignment. Not much else to say here.
  - Spaces around all types of operators, except for concatenation (`.`), object access (`->`), unary (`++`, `--`, `!`, etc.) and ternary in case the "if" is empty (?:)
  - Spaces after commas, semicolons and colons (except colons before return types, not sure about that one still...)
  - No spaces within brackets and parentheses
  - Egyptian braces (opening brace on the same line as the statement) ALWAYS, even if it's a one-liner.
  - Parentheses to clarify expressions even if not necessary
  - Short style arrays (`[]`)
  - Single quotes (`'`) for pure strings, double quotes (`"`) in case code needs to run through the parser (e.g. using special characters `"\r\n"`). Variable interpolation in double quotes is discouraged, use concatenation or `sprintf()` instead.
  - Blank lines before control structures 
  - NO YODA - if you feel the need to write Yoda conditions in order to prevent bugs you may also want to rethink the way you write code.
  - Soft line length of 130 characters (GitHub repo view width)
  - Inline comments start with a double slash (`// comment`), PHPDoc blocks with a slash and double asterisk (`/** doc block */`), single lines of code are commented with a caret (`# foo();`), multiple lines of code are commented with a slash and single asterisk (`/* code block */`).
  - PHPDoc comments are encouraged, but not always necessary in code that's written for php 7.4+ and is using typed properties etc., However, if the property or parameter is ambiguous (`array`), a doc comment with a proper type should be added if possible (`/** @var int[] */`).
  - Classes, methods and variables should be descriptive, abstract classes, interfaces and traits should be named as such (reminder that there are no contracts, it's not a business here and also we're no artisans). Class naming should also follow the [PSR-4 autoloading standard](https://www.php-fig.org/psr/psr-4/).

As you can see, a lot of it aligns with the PSRs, but i find it more intuitive that way.

**Comment by codemasher** [comment]


**Reply by l-Rose-l** [reply]
So your saying even if I had no experience in coding you would allow me to help lol?

**Reply by codemasher** [reply]
> 
> 
> So your saying even if I had no experience in coding you would allow me to help lol?

Why should I stop you from helping? Maybe you have a great idea to contribute regardless?
I've seen projects with horrible code but a 3 paragraph explanation in the readme why it's crucial to adhere to a certain coding standard - this is a form of gatekeeping i don't want to engage in.
